--- /volumio/app/musiclibrary.js.orig	2019-08-09 13:13:17.000000000 +0000
+++ /volumio/app/musiclibrary.js	2019-08-09 13:00:28.000000000 +0000
@@ -515,6 +515,9 @@
 			case 'Last_100':
 				self.browseSources[i].name = self.commandRouter.getI18nString('COMMON.LAST_100');
 				break;
+			case 'New_50':
+				self.browseSources[i].name = self.commandRouter.getI18nString('COMMON.NEW_50');
+				break;
 			default:
 			console.log('Cannot find translation for source'+self.browseSources[i].name)
 		}
